core/src/main/java/org/infinispan/persistence/file/SingleFileStore.java

```diff
@@ -127,76 +139,86 @@
+   private long filePos;
    private File file;
    private float fragmentationFactor = .75f;
    // Prevent clear() from truncating the file after a write() allocated the entry but before it wrote the data
    private final StampedLock resizeLock = new StampedLock();
    private TimeService timeService;
    private MarshallableEntryFactory<K, V> entryFactory;
-
-   public static File getStoreFile(GlobalConfiguration globalConfiguration, String locationPath, String cacheName) {
-      Path location = PersistenceUtil.getLocation(globalConfiguration, locationPath);
-      return new File(location.toFile(), cacheName + ".dat");
+   private KeyPartitioner keyPartitioner;
+   private BlockingManager blockingManager;
+   private boolean segmented;
+   private int actualNumSegments;
+   private int maxEntries;
+
+   public static File getStoreFile(String directoryPath, String cacheName) {
+      return new File(new File(directoryPath), cacheName + ".dat");
    }
 
    @Override
-   public void init(InitializationContext ctx) {
+   public CompletionStage<Void> start(InitializationContext ctx) {
       this.ctx = ctx;
       this.configuration = ctx.getConfiguration();
       this.timeService = ctx.getTimeService();
       this.entryFactory = ctx.getMarshallableEntryFactory();
+      this.blockingManager = ctx.getBlockingManager();
+
+      keyPartitioner = ctx.getKeyPartitioner();
+      maxEntries = configuration.maxEntries();
+      segmented = configuration.segmented();
+      if (segmented) {
+         actualNumSegments = ctx.getCache().getCacheConfiguration().clustering().hash().numSegments();
+      } else {
+         actualNumSegments = 1;
+      }
+
+      entries = new Map[actualNumSegments];
+      freeList = Collections.synchronizedSortedSet(new TreeSet<>());
+
+      // Not really blocking because no other thread can access the lock during start
+      blockingAddSegments(IntSets.immutableRangeSet(actualNumSegments));
+      return blockingManager.runBlocking(this::blockingStart, "sfs-start");
    }
 
-   @Override
-   public void start() {
-      try {
-         file = getStoreFile(ctx.getGlobalConfiguration(), configuration.location(), ctx.getCache().getName());
-         if (!SecurityActions.fileExists(file)) {
-            if (configuration.ignoreModifications()) {
-               entries = Collections.emptyMap();
-               return;
-            }
-            File dir = file.getParentFile();
-            if (!SecurityActions.createDirectoryIfNeeded(dir)) {
-               throw PERSISTENCE.directoryCannotBeCreated(dir.getAbsolutePath());
-            }
-         }
-         channel = SecurityActions.openFileChannel(file);
+   private void blockingStart() {
+      boolean shouldClear = configuration.purgeOnStartup();
+      boolean readOnly = configuration.ignoreModifications();
+      assert !(shouldClear && readOnly) : "Store can't be configured with both purge and ignore modifications";
```

Comment:
We should probably have this in a generic spot so that all stores can benefit from it.

Reply:
`AbstractStoreConfigurationBuilder.validateStoreAttributes()` already throws an exception for this combination, I just wanted to be extra sure :)